### PR TITLE
Add parquet SQL benchmarks

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #TBD.

 # Rationale for this change

Benchmarks good, more benchmarks more good :smile: 

# What changes are included in this PR?

This adds a benchmark that optionally generates a large-ish parquet file, or uses a file specified by an environment variable, and then runs through a list of queries against this file.

My hope is that this will supplement the TPCH benchmark, with one that is perhaps easier for people to setup and run, and that can be more easily adapted to test different data shapes and queries.

In particular as currently configured this will test:

* Dictionary arrays
* Nullable arrays
* Large-ish parquet files (~200Mb)
* Basic table scans with filters and aggregates
* ...Suggestions welcome :smile: 

It could theoretically be extended to incorporate joins, however, as I don't currently have a real-world use-case that produces these, I'd rather leave this to someone with such a workload to model a representative benchmark for.

_Unfortunately the generation portion needs https://github.com/apache/arrow-rs/pull/1214 but arrow 9 should be out soon which will contain this. Will keep this as a draft until then._

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by tustvold** [comment]
There are definitely tweaks that would be cool to make to this, e.g. testing different column encodings, but I think this is a decent starting point

File: dev/release/rat_exclude_files.txt [hunk 1]
```diff
@@ -116,6 +116,7 @@ ci/*
 **/*.svg
 **/*.csv
 **/*.json
+**/*.sql
```
Comment (tustvold) [inline dev/release/rat_exclude_files.txt:119]: I _think_ this is the correct thing to do, but someone should probably verify if RAT is needed for SQL files

Comment (alamb) [reply dev/release/rat_exclude_files.txt:119]: This is fine in my opinion

**Review by alamb — APPROVED** [review]
I also ran this locally 👌  very nice:

```shell
cargo bench --bench parquet_query_sql
...
Generating parquet file - /var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/parquet_query_sqlr7Ymzm.parquet
Generated parquet file in 6.7890725 seconds
Using parquet file /var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/parquet_query_sqlr7Ymzm.parquet


...
ng select dict_10_required, dict_100_required, MIN(f64_optional), MAX(f64_optional), AVG(f64_optional) ...: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.9s, or reduce sample count to 30.
Benchmarking select dict_10_required, dict_100_required, MIN(f64_optional), MAX(f64_optional), AVG(f64_optional) ...: Collecting 100 samples in estima                                                                                                                                                      select dict_10_required, dict_100_required, MIN(f64_optional), MAX(f64_optional), AVG(f64_optional) ...                        
                        time:   [128.35 ms 128.65 ms 128.98 ms]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe


```

@Igosuki  this might be a cool thing to run on the arrow2 branch to see how the performance compares

**Comment by alamb** [comment]
Looks like there is also a clippy complaint here

**Comment by Igosuki** [comment]
> I also ran this locally ok_hand very nice:
> 
> ```shell
> cargo bench --bench parquet_query_sql
> ...
> Generating parquet file - /var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/parquet_query_sqlr7Ymzm.parquet
> Generated parquet file in 6.7890725 seconds
> Using parquet file /var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/parquet_query_sqlr7Ymzm.parquet
> 
> 
> ...
> ng select dict_10_required, dict_100_required, MIN(f64_optional), MAX(f64_optional), AVG(f64_optional) ...: Warming up for 3.0000 s
> Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.9s, or reduce sample count to 30.
> Benchmarking select dict_10_required, dict_100_required, MIN(f64_optional), MAX(f64_optional), AVG(f64_optional) ...: Collecting 100 samples in estima                                                                                                                                                      select dict_10_required, dict_100_required, MIN(f64_optional), MAX(f64_optional), AVG(f64_optional) ...                        
>                         time:   [128.35 ms 128.65 ms 128.98 ms]
> Found 5 outliers among 100 measurements (5.00%)
>   3 (3.00%) high mild
>   2 (2.00%) high severe
> ```
> 
> @Igosuki this might be a cool thing to run on the arrow2 branch to see how the performance compares

I will rebase once it is merged

**Comment by alamb** [comment]
I will sort out the clippy complaint

File: datafusion/benches/parquet_query_sql.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                    let query = context.sql(query).await.unwrap();
+                    let mut stream = query.execute_stream().await.unwrap();
+                    while let Some(_) = criterion::black_box(stream.next().await) {}
+                })
+            });
+        });
+    }
+
+    // Clean up temporary file if any
+    std::mem::drop(temp_file);
```
[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/benches/parquet_query_sql.rs:233]: It was intended as a hint that the lifetime of `temp_file` matters, i.e. it must live to the end of the benchmark block. In the past I've accidentally refactored tests with NamedTempFile and its broken in odd ways that have boiled down to the temporary file getting cleaned up too early

**Pull request merged**